### DISCUSSION TITLE
Feature request: An "undefined" value in addition to "null"

### DISCUSSION THREAD

**Discussion by vhe1** [dotnet/csharplang#385] [General]
As web backend developer, the thing that is hardest for me is to deal with the two ways javascript frontends have to do nulls, undefined (member not in json) and null (member in json and has null assigned).

I'm constantly in discussions with my client developers and this is no fun at all.

The typical usage is described in [rfc 7386](https://tools.ietf.org/html/rfc7386), it's obvious, sensible and there is no way at all for us in the backend to deal with it because we only have null.

Of course, we could deserialize the json into Dynamic or a set of nested Dictionary<String,Object> structures but that would lose all type safety and convenience. Then we might as well use Typescript ([which has it](https://www.typescriptlang.org/docs/handbook/basic-types.html))  and node.js for backend programming.

This is the biggest gripe I currently have.

So, is there a way to allow us backend developers to correctly derive the client's intent from a json with nulls and missing members?

Maybe another method for ValueType would do it for the simple stuff? No idea about references though. On the other hand, C++ has so many pointers now, maybe c# can stand another reference type?

**Comment by MgSam** [comment]
This sounds like a serialization issue, not a language issue. As you yourself say, it's horrible that JavaScript has two ways of saying `null`, so we wouldn't want to propagate that mistake to other languages.

**Comment by Pzixel** [comment]
@vhe1 null itself is a billion-dollar mistake and you want to just square it. It's JS problems that it has two entities to describe the almost same thing - absence of something. In my practice no one cares if value is not setted at all or it's just an unusable value. If field is required - you are rising an error, if it isn't then both are valid. But you have a lot of problems with `passing null or undefined` guidelines or comparing with `absence`... It's a no-no feature, don't even think about it.

**Comment by vhe1** [comment]
I did not say that "it's horrible that JavaScript has two ways of saying null", I said it's horrible that we can't use the information contained in this in a typed backend.

Re "null is a billion dollar mistake", that's just wrong. The problem is not that there is null but that there is just one of it, forced to express two different intents. (Please remember the discussons in SQL. They focused precisely on the fact that it was not clear _which_ null.) Having two different nulls solves that.

**Comment by svick** [comment]
@Pzixel 

> In my practice no one cares if value is not setted at all or it's just an unusable value. If field is required - you are rising an error, if it isn't then both are valid. But you have a lot of problems with passing null or undefined guidelines or comparing with absence... It's a no-no feature, don't even think about it.

@vhe1 links to an RFC for JSON Merge Patch, which is an actual specification that differentiates between undefined and null. So it may not be a choice to "not think about it", if you need to interact with a system that uses that format.

**Comment by HaloFour** [comment]
`Optional<T>`.  Done.

**Comment by svick** [comment]
@vhe1 The problem with `null` in a statically typed language like C# is that you don't have a choice about using it, it's always a valid value. C# is trying to move away from that (see https://github.com/dotnet/csharplang/issues/36). Adding a second value that is valid for all reference types would be a move in the other direction and, in my opinion, a bad idea.

I think what you should do is to use the type system to your advantage: instead of a property of some type `T`, use a property of type, say, `Value<T>`, whose API looks something like this:

```c#
public struct Value<T>
{
    public static Value<T> Undefined { get; }
    public static Value<T> Null { get; }
    public static Value<T> Create(T value); // throws if value is null

    public bool IsUndefined { get; }
    public bool IsNull { get; }
    public T Value { get; } // throws exception when IsUndefined == true or IsNull == true
}
```

Then you teach your JSON serializer about `Value<T>`, and you're set. This has the advantages that it limits `undefined` to your serialization code and makes it very clear that `undefined` (and `null`) is a possibility.

There is no need to pollute all of C# with this.

**Comment by vhe1** [comment]
I've tried stuff along those lines but either it looks horrible in any swagger definition or it has a hole somewhere else.
Maybe it's also telling that there is no nuget package for this yet.

But I'll try some more then.

Thanks a lot for bothering!

**Comment by aluanhaddad** [comment]
```C#
Newtonsoft.Json.NullValueHandling.Ignore
```

**Comment by vhe1** [comment]
aluanhadad: How does this help me during deserialization?
The point is that I _get_ json like this and want to extract the null/undef information from it.

**Comment by aluanhaddad** [comment]
@vhe1 1

I'm may have written something a bit wrong below, I am really tired 😪

So the simple scenario is C# producer, using Newtonsoft (the JSON one, not the [DSON](https://vpzomtrrfrt.github.io/DSON/) one), paired with a JavaScript consumer using `JSON.parse` and `JSON.stringify`.

To do this, we need to form a semantic bridge, across JSON, from C# types (CLR) to JavaScript values (ECMAScript) types. JSON has structural types only. There are no named constructs, it is nice and simple. It also happens to be a rough subset of ECMAScript literals which makes this mapping relatively straightforward.

The idea is basically to adopt the view that `null` in C# maps to `undefined` in JavaScript.<sup>1</sup>
This means that `null` is not serialized into any of your responses and that it is not plucked from any of your requests.

We can enforce this as follows:
C#
```C#
class Person
{
    public string Name { get; set; }
    public int? Age { get; se; }
}

var person = new Person
{ 
    Name = "Anders"
}; // Age is null
```
JSON
```JSON
{
  "name" : "Anders"
}
```
JavaScript 
```js
// literal from JSON.parse(response.data)
const person = { 
  name: 'Anders'
};
```
So we don't know the age of our person on either side and our REST interface allows it to be _unspecified_. So all is well.

Now when the JavaScript client wants to send me some values, he can do anything he wants
1. Nice JavaScript approach
```js
const person = {
  name: 'Douglas'
};

request.data = JSON.stringify(person); // JSON => { "name": "Douglas" };
```
2. Ugly JavaScript approach (bad style)
```js
const person = { 
  name: 'Douglas',
  age: null
};

request.data = JSON.stringify(person); // JSON => { "name": "Douglas", "age": null };
```
Thanks to `Newtonsoft.Json.NullValueHandling.Ignore`, the C# code on the server will simply ignore the difference between `null` and `undefined` (missing) that the client produces. This works partially because I made `Age` of type `int?` but you have to in order to map these correctly regardless of how you handle serialization.

C#
```C#
public HttpResponseMessage Post(Person person)
{
    var person = JsonConvert.DeserializeObject<Person>(Model);
    Console.WriteLine(person.Name); // => "Douglas"
    Console.WriteLine(person.Age); // => null
}
```

**(1)** This applies equally to TypeScript (obviously as TS is JS). Interestingly, while the community is somewhat divided, a fair number of ECMAScript experts such as Douglas Crockford, as well as the TypeScript team, have stated that `null` is unnecessary and have removed it from their code bases, include the TypeScript compiler.

**Comment by vhe1** [comment]
You are still trying to solve the wrong problem.

The problem I want to solve is to READ a json object according to rfc 7386 as I wrote in my initial feature request posting.

I'm not interested in creating such a json, I want to READ it and in the generated c# object distinguish between JSON null and JSON undefined, in order to properly deal with it according to that rfc.

So, according to rfc 7386, {name: "XXX", age: null} and {name:"XXX"} have two distinct meanings. The first one updates the name and /deletes the age from whatever persistence layer there is/, the second one updates the name and /leaves the age alone/.

**Comment by Pzixel** [comment]
@vhe1 then you just should write your own deserializer and wrap every propery with `Option<T>`. C# shouldn't implement such feature because there is some language implementing it. It easily can be done if serializer distinguish between these two states. And probably JSON.Net is following it when using `JObject.Parse`. The problem is that `JavaScript` (and thus JSON specification) mix `dictionary` and ordinal `object` type. In C# you just cannot have `undefined` - all objects of class `A` always has all properties. On the other hand we have `Dictionary<string, object>` and you always know when value is not present in dictionary (`ContainsKey -> false`) or value is null (`foo["bar"] == null`).

**Comment by vhe1** [comment]
@Pzixel, I understand that, I was just replying to aluanhaddad.

**Comment by aluanhaddad** [comment]
@vhe1 I misunderstood your use case. You are dealing with partial updates. In that case, I think the solution is indeed what @HaloFour and @Pzixel are suggesting.

You can wrap your fields in an optional value. You can configure your serializer to understand it, which is not hard.

The `Option<T>` type itself is trivial to write.

**Comment by john-cd** [comment]
Sidebar: the R language has both NULL and NA. Missing values are represented by the symbol NA (not available).  

Having NA is really useful in stats, data science, etc...

**Comment by aluanhaddad** [comment]
> Sidebar: the R language has both NULL and NA. Missing values are represented by the symbol NA (not available).

> Having NA is really useful in stats, data science, etc...

But, do you ever have a dataset containing both `NULL` and `NA`?

**Comment by vhe1** [comment]
(I understand that I should try to get an Option<T> working, so I'm responding now just to the question above.)

In JSON you model `NA` by leaving out the property altogether. So, in transit and in most NoSQL Databases you can have two fields, one existing but having the `null` token as value, and one not existing and there are ways to distinguish between the two.

**Comment by JoergWMittag** [comment]
I don't think this is a good idea. There are just so many different scenarios that one might want to cover that it is impossible to handle them all:

1. We know the value.
2. We know that the value doesn't exist.
3. We know that the value does exist, but we don't know what the value is.
4. We don't know whether the value exists or not.
5. We know that the value exists, and we know how to get the value, but we didn't get it yet.
6. … And many others that I can't think of right now …

All of these can already be handled by defining suitable generic types, e.g. 5) is a `Task<T>`, 4) is a `Maybe<T>` and so on.

I once read somewhere that the database guys have now identified over a dozen different kinds of `null`s, do you really want to add all of them to C♯?

**Comment by AustinBryan** [comment]
C# is trying really, really hard to move away from `null` values, and you want another one? 😂

**Comment by vdachev** [comment]
I believe `null` and `undefined` values have different semantic meaning.  For example, when representing a nodes in a tree, you might set the `Parent` to `null` to indicate that the node has no parent and `undefined` to indicate that it might have a (`null`) parent but it's not set (see below).

Here are some additional scenarios to illustrate the need for an `undefined` value:

- **Search**. Consider the following snippets:
  ```csharp
  class Item {
    public int ParentId { get; set; }
  }

  class ItemFilter {
    public int? ParentId { get; set; }
  }
  ```
  How do you differentiate between the following two situations?
    1. The consumer needs all the nodes of the tree (the items, regardless of the value of their `ParentId` property).
    2. The consumer needs the root nodes of the tree (the items, having the value of their `ParentId` property set to `null`).
- **Updating**.  Consider the following snippet:
  ```csharp
  class Item {
    public int ParentId { get; set; }
  }

  class UpdateItemRequest {
    public int? ParentId { get; set; }
  }
  ```
  How do you differentiate between the following two situations?
    1. The consumer doesn't want to change the value of the `ParentId` property of the item.
    2. The consumer wants to set the value of the `ParentId` property of the item to `null`.
- **NoSQL** (i.e. MongoDB). How do you tell if a property of a document exists or set to `null`?

IMHO, such scenarios are so common that it's at least worth considering adding some syntactic sugar in C# around a type similar to the `Value<T>` suggested above.

**Comment by svick** [comment]
@vdachev 

> IMHO, such scenarios are so common that it's at least worth considering adding some syntactic sugar in C# around a type similar to the `Value<T>` suggested above.

I'm not sure that syntactic sugar is necessary. If you think this type would be useful, it might make sense to create an issue on [the corefx repo](https://github.com/dotnet/corefx), so that the type is added to .Net.

**Comment by sharwell** [comment]
Assuming the underlying issue here is situations like https://github.com/openstacknetsdk/openstack.net/issues/445, I think I understand how a request like this comes into existence. However, I believe approaches like Roslyn's [`Optional<T>`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Optional.cs) (which is *very* different from "maybe" types) provide a better solution to this domain-specific problem.

**Comment by DarthVella** [comment]
@vdachev 

Forgive my bluntness, but I'd much prefer explicit, automatically documented behaviour rather than guessing which of null or undefined the producer decided should do what.

**Search**
```csharp
class ItemFilter {
  public int? ParentId { get; set; }
  public bool RootNodesOnly { get; set; }
}
```

**Updating**
```csharp
class UpdateItemRequest {
  public Dictionary<string, object> UpdateValues { get; set; }
}
//Alternately
class UpdateItemRequest {
  public int? ParentId { get; set; }
  public string[] ParamsToUpdate { get; set; }
}
```

**Comment by vdachev** [comment]
@svick , we already have similar syntactic sugar around `INullable`, as well as around other common scenarios such as `IDisposable`, `IEnumerable`, `Task`, etc.

@sharwell , as I have illustrated above, it's not just JSON-related scenarios. It's a matter of providing a convenient API to work with.

@DarthVella , I wouldn't resort to these solutions because:
  - In the **Search** scenario, what if the consumer both sets `ParentId` to a non-null value and `RootNodesOnly` to `true`? These are mutually exclusive and it should be handled this scenario during runtime. Also, this would help in the case of a tree. How about any other scenarios where one should be able to filter objects either by `null`, or regardless of the value? Would you add a second property for each nullable?
  - In the **Update** scenario, you allow the consumer to use arbitrary property name (and multiple properties by the same name might exist in the hierarchy) and/or lose any type safety of the values. You could, or course, address the first issue using a `Expression<Func<T>>` and face the problem with their serialization.

**Comment by DarthVella** [comment]
I agree that those solutions have problems to overcome, of course; but they are, in my opinion, clearer issues with straightforward answers or known alternatives.

**Comment by PaulGarnerUk** [comment]
I too have been battling with this the last few days, exactly as the original questioner mentioned - trying to implement [merge-patch+json](https://tools.ietf.org/html/rfc7386)

I think a lot of the early comments here were missing the point, and I too have to admit that having fallen in love with Typescript recently I'd never have imagined a scenario where I wanted both null and undefined again!  But accepting merge-patch+json requests is exactly the use case.

I've currently adapted and extended the [`Optional<T>`](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Optional.cs) from Roslyn and have wrapped all my properties, but it has not been trivial.

Things like reflecting out the values from the properties, json de/serialization, and diddling Swagger to hide the `Optional<>` and instead show the underlying type all required quite a bit of head scratching and jumping through hoops, and I'm sure there will be further pain points as I continue. I'd be confident the C# team could take a lot of this pain away if there was a framework implementation that covered all the corner cases, and I'd love to see someone take this on

**Comment by CyrusNajmabadi** [comment]
> I'd be confident the C# team could take a lot of this pain away if there was a framework implementation that covered all the corner cases, and I'd love to see someone take this on

This sounds liek a request for a framework/api to help out here.  It doesn't seem like it would need any particular language support.

> I think a lot of the early comments here were missing the point, and I too have to admit that having fallen in love with Typescript recently I'd never have imagined a scenario where I wanted both null and undefined again! But accepting json-merge+patch requests is exactly the use case.

Can you provide more information?  Why does this require an undefined value to be provided at the *language* level?  Why can this not just be done at an API level?

**Comment by CyrusNajmabadi** [comment]
Note: i've read the entire tiny spec around RFC 7386.  I don't see anything about it that would be difficult to implement...

**Comment by PaulGarnerUk** [comment]
Fair point, because as has been discussed, I think it is possible to tackle it outside the language. My point was more to address that there are complexities that aren't immediately obvious until you've perhaps tried to implement [merge-patch+json](https://tools.ietf.org/html/rfc7386). 

For example my first attempt was based around [https://github.com/Morcatko/Morcatko.AspNetCore.JsonMergePatch](https://github.com/Morcatko/Morcatko.AspNetCore.JsonMergePatch). 

That basically builds a [JsonPatchDocument](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.jsonpatch.jsonpatchdocument?view=aspnetcore-2.1) during deserialization, and if you only want to receive a model and make updates from what was received - that works pretty well. But if you want to receive a model, then have some backend logic to (eg) uppercase one of the property values, or set/null a property before the update then you'd need to start modifying the underlying JsonPatchDocument too (with it's quirky syntax which looks like it's more about supporting 'the other' JSON patch standard - [RFC6902](https://tools.ietf.org/html/rfc6902)

I gave up on that and instead started over with this idea of wrapping the properties with `Optional<T>`. I'm sure developers far better than me might find it trivial, but I personally found writing a de/serializer to handle (eg) `Optional<Nullable<Enum T>>` was not a simple task, and I expect there will be other things that prove similarly challenging as I continue to use it.

My point I suppose is to agree with you that yes, this can be tackled outside the language. But if merge-patch+json becomes popular (and in my opinion it's the most natural way for clients to submit JSON updates), then language support would make it easier.

**Comment by Pzixel** [comment]
We'd better to not have nulls at all, and have `Nullable<T>` for both reference and value types in BCL, but unfortunately it's too late to change things.

However, adding `undefined` into the language would make almost all other scenarios a nightmare.

> My point I suppose is to agree with you that yes, this can be tackled outside the language. But if merge-patch+json becomes popular (and in my opinion it's the most natural way for clients to submit JSON updates), then language support would make it easier.

Do you want XML or TOML support built in the language as well?

**Comment by CyrusNajmabadi** [comment]
@PaulGarnerUk Can you provide more details?  None of what you just mentioned makes me feel like the language needs an 'undefined' concept.

**Comment by vdachev** [comment]
Okay, as several people and I have already mentioned - there are multiple common scenarios with updating and searching of objects. Translating the notion of `undefined` from JSON to C# and vice versa requires significant effort, just because C# does not support this notion. But it's not just JSON.

There's also XML serialization. An XSD may contain a declaration such as:

```xml
<xs:element minOccurs="0" name="Name" nillable="true" />
```

Because of this particular languge limitation, tools like "xsd.exe" generate an additional `public bool *Specified { get; set; }` property next.

Also, SQL allows you to select, insert and update rows with values for selected columns only, NoSQL enables you to patch objects, too. How do I provide a convenient plain C# API for doing so if I can't have decent search/create/update request type?

There are numerous scenarios that such language feature would be an elegant (and uniform) solution to.

And honestly, I'm disappointed that the counterarguments lack here are limited to statements such as "I'm still not convinced." and "There are other ways to achieve this." without any further details.

**Comment by theunrepentantgeek** [comment]
>  I'm disappointed that the counterarguments lack here are limited to statements such as "I'm still not convinced." and "There are other ways to achieve this." without any further details.

But that's not how this works.

You're wanting a language change that will impact on every user of the language, for as long as there are developers maintaining systems written in C#. That's tens of millions of current developers, and at least twenty years.

Implementation of any language change requires substantial planning and design, to ensure that it interacts in useful/predictable ways with every other feature that already exists. This is a substantial piece of work.

Every suggested feature [starts at -100 points](https://blogs.msdn.microsoft.com/ericgu/2004/01/12/minus-100-points/) and needs to offer substantial value just to make it to a positive score where it's worth considering.

My problem with this suggestion is that it adds another **null**-like value into the language and type system, multiplying the ways that things can go wrong. 

[Tony Hoare](https://en.wikipedia.org/wiki/Tony_Hoare) _invented_ **null** and describes it as his ["Billion dollar mistake"](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare). I've read articles in [CACM](https://cacm.acm.org/) estimating that the cost to the _US economy alone_ exceeds a billion $US per year.

With [Nullable Reference types](https://github.com/dotnet/csharplang/issues/36) the folks behind C# are doing a very large amount of work to try and mitigate these costs, making it much easier for the rest of us to identify and remedy **null** reference issues as early as possible. They've put a lot of work into making the impact as positive as possible, but there will be costs for all of us as we migrate forward and address the lurking issues revealed in the code we already have.

Not to mention a number of JavaScript and Typescript developers I know who bemoan the presence of **undefined** in those languages because of all the grief it causes.

Given all the effort/cost/time that's being invested in reducing the impact of **null**, any proposal to introduce **undefined** (or anything else of that ilk) into C# is going to need to include a really compelling set of benefits - something more than addressing some edge cases for serialization at the borders of our systems.

So far, it seems to me that better tooling is extremely likely to be sufficient. 

Convince me to change my mind. 😀 

(Perhaps a good solution would be something that moves away from POCOs in the style of [The Best ORM you've never used](http://www.nichesoftware.co.nz/2012/08/18/the-best-orm-that-youve-never-used.html) Disclaimer: That's a link to an old post on my blog.)

**Comment by orthoxerox** [comment]
I wouldn't be against supporting `undefined` as a singleton type when (if) we get union types. Then JSON values could be encoded as `string | double | bool | Null | Undefined | List<JsonValue> | Dictionary<string, JsonValue>`.

**Comment by CyrusNajmabadi** [comment]
> And honestly, I'm disappointed that the counterarguments lack here are limited to statements such as "I'm still not convinced." and "There are other ways to achieve this." without any further details.

It's not a counter argument.  It's literally: there hasn't been a justification given why this is needed.  For example, on several occasions i responded to someone stating that this would be helpful for json-merge-patch.  I inquired to find out *how* this would be helpful, but no answer was forthcoming.

If the actual use case cannot be explained, then it's going to be extremely difficult to convince someone that this should be added to the language.

**Comment by CyrusNajmabadi** [comment]
> Okay, as several people and I have already mentioned - there are multiple common scenarios with updating and searching of objects.

Ok... but what does that mean?  Why does C# need an 'undefined' value to help you search/update objects?

Another way to date this: i have written code for the last 20 years in C# that searches/updates objects.  So far, this hasn't been problematic.  So it's not immediately obvious what problem this is actually addressing.  If you think tehre is a problem, please explain it *so i can understand it*.

> Also, SQL allows you to select, insert and update rows with values for selected columns only, NoSQL enables you to patch objects, too. How do I provide a convenient plain C# API for doing so if I can't have decent search/create/update request type?

How do the existing Nosql/Sql libraries currently let you do this? 

>  and "There are other ways to achieve this." without any further details.

It's not the job of others to do that for you :)   If you're the one asking that a new feature be added, the onus is on you to show why it's important enough.  By default, the standard position of the language is that it's going to stay as is unless strong reason otherwise.  That's the starting point.  People don't need to defend that as it's the basis for how the language is developed.

**Comment by PaulGarnerUk** [comment]
> on several occasions i responded to someone stating that this would be helpful for json-merge-patch. I inquired to find out _how_ this would be helpful, but no answer was forthcoming.

I'd only be repeating various posts above. With json-merge+patch when deserializing into a model that is (eg)

```
{
    public string Name { get; set; }
    public int? HouseNumber { get; set; }
}
```

If the client wants to change the house number to a value they send:
```
{
     "houseNumber": 123
}
```

If the client wants to change the house number to null they send:
```
{
    "houseNumber": null
}
```

If they don't want to make ANY CHANGE to the house number, but instead want to change the name, they send:
```
{
    "name": "Arthur Smith"
}
```

That means when we deserialize those three inputs, one has a value of `123`, one has a value of `null`, and the last has a value of `undefined`.  C# has no concept of undefined, so custom handling is required. That custom handling is not straightforward and needs adopting wholesale throughout the solution

Might be nice if C# had something natively that would allow a check like `houseNumber.HasValue` which if it returned `false` meant no value had been set at all (value is undefined). If it returned `true` it would mean a value has been set, and (if the type is nullable) that value could be explicitly null.

**Comment by Pzixel** [comment]
@YairHalberstadt I do believe you pas `"houseNumber": undefined`. However, I'm not sure how it's related to C#.

@PaulGarnerUk then just write
```cs
{
    public string Name { get; set; }
    public Option<int?> HouseNumber { get; set; }
}
```

The problem is that you should nest nullable in Option and it's not really convinient. It's because C# lacks of ADT, so it's more like XY problem, because you probably want to write

```cs
enum JsonValue<T> {
   Null,
   Undefined,
   Some(T)
}
...
{
    public string Name { get; set; }
    public JsonValue<int> HouseNumber { get; set; }
}
```

**Comment by jnm2** [comment]
@Pzixel 

` undefined` is an invalid JSON value according to the JSON spec. No well-formed JSON will contain it.

@PaulGarnerUk

If you don't know ahead of time which JSON attributes will exist, I'd say it's more useful to inspect a JSON DOM than to deserialize to a POCO.

**Comment by PaulGarnerUk** [comment]
@jnm2  
> If you don't know ahead of time which JSON attributes will exist, I'd say it's more useful to inspect a JSON DOM than to deserialize to a POCO

I don't understand how checking a DOM will help? We still need to deserialize to a C# object, and the C# object needs a property with three possible states: value specified, explicit null value, or unspecified value.

@Pzixel - I at least feel you understand the issue, thanks. 👍 The `Optional<T>` in Roslyn is useful, but it needs a custom de/serializer amongst other workarounds.  As I've said - this is the route I'm using to tackle it, but it hasn't been easy.

**Comment by YairHalberstadt** [comment]
How would you Serialise a C# object that had an undefined value for one of its properties?

You'd end up putting in a feature request to add undefined to Json. And then you'd need to add reallyReallyUndefined to C#. And the never ending cycle continues...

**Comment by HaloFour** [comment]
@PaulGarnerUk 

IMO, the real problem is that the BCL doesn't have a proper stock `Optional<T>` type.  If it did, all of the standard serializers could support it out of the box and you wouldn't have to do anything aside declaring your fields as that type.

**Comment by vdachev** [comment]
I've been writing C# code for the last 13 years and I've already mentioned several cases when an `undefined` would make life much easier. LINQ query syntax was not required, neither was the `?` for `INullable`, `using` for `IDisposable`... but they make C# development a charm.
I'm not saying the solution is necessarily introducing an `undefined` keyword. It could be some language syntax extension around `Optional`, or a feature like a union (as @orthoxerox suggested).

@theunrepentantgeek , does ACM advise against the very existance of `null` values? Nope, and the C# team makes an effort in 8.0 to add safety checks, not to scap `null` altogether. JavaScript lacks those safety checks and that's the reason `undefined` is an issue. TypeScript has the language features to enable developers to avoid problems with both `null` and `undefined` inherited from JavaScript.

@CyrusNajmabadi , in SQL you can choose which values to include in an INSERT and UPDATE statements, and WHERE clauses. Entity Framework Core handles these scenarios with its own workaround - object tracking.

@YairHalberstadt ,
> How would you Serialise a C# object that had an undefined value for one of its properties?

The property will not be present in JSON - that's what `undefined` is for, unassigned, not set...

**Comment by PaulGarnerUk** [comment]
> How would you Serialise a C# object that had an undefined value for one of its properties?

Turtles all the way down!  No, by not specifying the property/value in JSON, then by definition it has an undefined value (neither a tangible value, nor null).  Undefined does not get serialized/deserialized, as opposed to null which, as you know, can be serialized.

**Comment by jnm2** [comment]
@PaulGarnerUk 

> We still need to deserialize to a C# object, and the C# object needs a property with three possible states: value specified, explicit null value, or unspecified value.

Why is this the only option? I can imagine taking code which uses the POCO and rewriting it to use the DOM. For something that has a very dynamic schema, this might even be preferable.

**Comment by CyrusNajmabadi** [comment]
It sounds like people are talking about an API they'd like, or the behavior of existing APIs.  I still haven't seen a use case fo C# having an actual undefined value.

**Comment by CyrusNajmabadi** [comment]
> but it needs a custom de/serializer amongst other workarounds

Wouldn't 'undefined' be the same way? All these tools and apis would have to be updated to handle this new intrinsic value.  If they can be updated to handle 'undefined', why can't they be updated to handle 'Optional'?

**Comment by theunrepentantgeek** [comment]
The crux of the problem seems to be the desire to have a distinct value (a sentinel value) that means *don't overwrite*.

To illustrate, you can use **null** as the sentinel value that means *keep the existing value*:

``` csharp
public class Person
{
    public string FullName { get; private set; }
    public string KnownAs { get; private set; }
    public string FamilyName { get; private set; }

    public void Patch(Person other)
    {
        if (other.FullName is not null) FullName = other.FullName;
        if (other.KnownAs is not null) KnownAs = other.KnownAs;
        if (other.FamilyName is not null) FamilyName = other.FamilyName;
    }
}
```

But that doesn't let you patch existing properties to **null**, because you're using **null** to indicate *don't overwrite*.

So you introduce a new sentinel value, let's say **undefined**.

``` csharp
public class Person
{
    public string FullName { get; private set; }
    public string KnownAs { get; private set; }
    public string FamilyName { get; private set; }

    public void Patch(Person other)
    {
        if (other.FullName is not undefined) FullName = other.FullName;
        if (other.KnownAs is not undefined) KnownAs = other.KnownAs;
        if (other.FamilyName is not undefined) FamilyName = other.FamilyName;
    }
}
```

The problem is, now you can't patch existing properties to **undefined**, because that's the sentinel value that indicates *don't overwrite*.

All you've done is to move the problem down the road one step.

The problem is caused by the use of a sentinel value instead of a distinct flag.

What I did, when I needed to solve this exact problem, was to create a **struct** that captured whether I'd specified a value or not. With epic amounts of imagination, I called it [`Specifiable<T>`](https://github.com/Azure/azure-batch-software-entitlement/blob/914abe4163ffc6b5bf6a176a60981e2f0366ee4a/src/Microsoft.Azure.Batch.SoftwareEntitlement.Common/Specifiable.cs).

**Comment by yaakov-h** [comment]
> The problem is, now you can't patch existing properties to undefined, because that's the sentinel value that indicates don't overwrite.

And with NoSQL systems, this is actually a perfectly valid use-case.

**Comment by vdachev** [comment]
> > but it needs a custom de/serializer amongst other workarounds
> 
> Wouldn't 'undefined' be the same way? All these tools and apis would have to be updated to handle this new intrinsic value. If they can be updated to handle 'undefined', why can't they be updated to handle 'Optional'?

So were all the API that use `Task`, `Nullable`, etc.

**Comment by CyrusNajmabadi** [comment]
@vdachev Exactly.  Task was added to the framework and the libraries updated to use it.  No need to make a language change here.  This should likely be moved to corefx with a request to make a suitable Optional/Some/None type for this purpose.

**Comment by vdachev** [comment]
> @vdachev Exactly. Task was added to the framework and the libraries updated to use it. No need to make a language change here. This should likely be moved to corefx with a request to make a suitable Optional/Some/None type for this purpose.

The point in this discussion is that there are lots of scenarios when an `undefined`-like feature would be extremely helpful. Every library/solution addresses the problem in a different way and once you get need to "bridge" between a few technologies (e.g. between MVC and Entity Framework Core) you essentially have to "translate" between them. A standard solution would certainly reduce the effort of developing these bridges.

Or course, it all depends on the way such support is implemented:

- One way is to get a decent `Option` support in BCL (.NET Standard?) first and then get syntactic sugar around it in the language (as @HaloFour suggested).
- Implement it as a language feature over CLR (or DLR).
- Something else, I can't think of now.

**Comment by CyrusNajmabadi** [comment]
> The point in this discussion is that there are lots of scenarios when an undefined-like feature would be extremely helpful. Every library/solution addresses the problem in a different way and once you get need to "bridge" between a few technologies (e.g. between MVC and Entity Framework Core) you essentially have to "translate" between them. A standard solution would certainly reduce the effort of developing these bridges.

As mentioned, the standard solution would be from Corefx, not from C#.  C# is not the standard.  It's one of many languages here.  You want a corefx solution so that tehse libraries can actually adopt it.  I recommend you open a bug there.   If tehre is some library that is created, and it becomes popular enough, there can be consideration around if there should be a language feature.

> Implement it as a language feature, if the CLR (or DLR).

I don't know what this means.

**Comment by jnm2** [comment]
`DBNull.Value` is crying and wondering why everyone is looking for a new sentinel value.
